### PR TITLE
JAMES-3942 Audit user trail

### PR DESCRIPTION
**Pull request by quantranhong1999**
resolve https://github.com/linagora/james-project/issues/4866

### CODE REVIEW

File: protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/esmtp/AuthCmdHandler.java [hunk 1]
```diff
@@ -387,6 +389,14 @@ protected Response doAuthTest(SMTPSession session, Username username, String pas
...
                         // TODO: Make this string a more useful debug message
                         AUTHENTICATION_DEDICATED_LOGGER.debug("AUTH method {} succeeded", authType);
+
+                        AuditTrail.entry()
+                            .username(username.asString())
+                            .remoteIP(session.getRemoteAddress().getAddress().getHostAddress())
+                            .protocol("SMTP")
+                            .action("AUTH")
+                            .parameters(ImmutableMap.of("authType", authType))
+                            .log("SMTP Authentication succeeded.");
```
Comment (chibenwa) [inline protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/esmtp/AuthCmdHandler.java:399]: Handle the fact that session.getRemoteAddress can be null (if the client disconnects)

File: server/container/util/src/main/java/org/apache/james/util/AuditTrail.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+            return this;
+        }
+
+        public Entry parameters(Map<String, String> parameters) {
+            this.parameters = parameters;
+            return this;
+        }
+
+        public void log(String message) {
+            MDCStructuredLogger.forLogger(LOGGER)
```
Comment (chibenwa) [inline server/container/util/src/main/java/org/apache/james/util/AuditTrail.java:81]: Do that only if logging is turned on please.

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/LocalDelivery.java [hunk 3]
```diff
@@ -62,6 +66,15 @@ public LocalDelivery(UsersRepository usersRepository, @Named("mailboxmanager") M
...
     @Override
     public void service(Mail mail) throws MessagingException {
+        AuditTrail.entry()
+            .protocol("mailetcontainer")
+            .action("LocalDelivery")
+            .parameters(ImmutableMap.of("mailId", mail.getName(),
+                "mimeMessageId", mail.getMessage().getMessageID(),
+                "sender", mail.getMaybeSender().asString(),
+                "recipients", StringUtils.join(mail.getRecipients())))
+            .log("Local delivering mail.");
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/LocalDelivery.java:76]: Do it one per recipient, straight after the per recipient success.

File: server/data/data-file/src/main/java/org/apache/james/mailrepository/file/FileMailRepository.java [hunk 4]
```diff
@@ -171,6 +174,15 @@ public void init() throws Exception {
...
     @Override
     public MailKey store(Mail mc) throws MessagingException {
+        AuditTrail.entry()
+            .protocol("mailrepository")
+            .action("store")
+            .parameters(ImmutableMap.of("mailId", mc.getName(),
+                "mimeMessageId", mc.getMessage().getMessageID(),
+                "sender", mc.getMaybeSender().asString(),
+                "recipients", StringUtils.join(mc.getRecipients())))
+            .log("FileMailRepository is storing mail.");
```
Comment (chibenwa) [inline server/data/data-file/src/main/java/org/apache/james/mailrepository/file/FileMailRepository.java:184]: Please log once the action is done, not before!

File: protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/esmtp/AuthCmdHandler.java [hunk 5]
```diff
@@ -387,6 +389,14 @@ protected Response doAuthTest(SMTPSession session, Username username, String pas
                     } else if (SMTPRetCode.AUTH_OK.equals(res.getRetCode())) {
                         // TODO: Make this string a more useful debug message
                         AUTHENTICATION_DEDICATED_LOGGER.debug("AUTH method {} succeeded", authType);
+
+                        AuditTrail.entry()
+                            .username(username.asString())
+                            .remoteIP(session.getRemoteAddress().getAddress().getHostAddress())
```
Comment (chibenwa) [inline protocols/smtp/src/main/java/org/apache/james/protocols/smtp/core/esmtp/AuthCmdHandler.java:395]: Let's add a field sessionId

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/LocalDelivery.java (hunk 3 above)
Comment (chibenwa) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/LocalDelivery.java:76]: Let's add a field: mailboxSessionId

File: server/mailrepository/mailrepository-blob/src/main/scala/org/apache/james/mailrepository/blob/BlobMailRepository.scala [hunk 6]
```diff
@@ -147,6 +145,15 @@ class BlobMailRepository(val blobStore: BlobStoreDAO,
...
   @throws[MessagingException]
   override def store(mc: Mail): MailKey = {
+    AuditTrail.entry
+      .protocol("mailrepository")
+      .action("store")
+      .parameters(ImmutableMap.of("mailId", mc.getName,
+        "mimeMessageId", mc.getMessage.getMessageID,
+        "sender", mc.getMaybeSender.asString,
+        "recipients", StringUtils.join(mc.getRecipients)))
+      .log("BlobMailRepository is storing mail.")
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-blob/src/main/scala/org/apache/james/mailrepository/blob/BlobMailRepository.scala:155]: Do this once the action succeeded.

File: server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepository.java [hunk 7]
```diff
@@ -64,6 +67,15 @@ public class CassandraMailRepository implements MailRepository {
...
     @Override
     public MailKey store(Mail mail) throws MessagingException {
+        AuditTrail.entry()
+            .protocol("mailrepository")
+            .action("store")
+            .parameters(ImmutableMap.of("mailId", mail.getName(),
+                "mimeMessageId", mail.getMessage().getMessageID(),
+                "sender", mail.getMaybeSender().asString(),
+                "recipients", StringUtils.join(mail.getRecipients())))
+            .log("CassandraMailRepository is storing mail.");
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-cassandra/src/main/java/org/apache/james/mailrepository/cassandra/CassandraMailRepository.java:77]: idem

File: server/mailrepository/mailrepository-memory/src/main/java/org/apache/james/mailrepository/memory/MemoryMailRepository.java [hunk 8]
```diff
@@ -38,7 +42,16 @@ public MemoryMailRepository() {
...
-    public MailKey store(Mail mail) {
+    public MailKey store(Mail mail) throws MessagingException {
+        AuditTrail.entry()
+            .protocol("mailrepository")
+            .action("store")
+            .parameters(ImmutableMap.of("mailId", mail.getName(),
+                "mimeMessageId", mail.getMessage().getMessageID(),
+                "sender", mail.getMaybeSender().asString(),
+                "recipients", StringUtils.join(mail.getRecipients())))
+            .log("MemoryMailRepository is storing mail.");
```
Comment (chibenwa) [inline server/mailrepository/mailrepository-memory/src/main/java/org/apache/james/mailrepository/memory/MemoryMailRepository.java:53]: idem

File: server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/SendMailHandler.java [hunk 9]
```diff
@@ -89,6 +92,18 @@ public HookResult onMessage(SMTPSession session, Mail mail) {
                         session.getRemoteAddress().getAddress(),
                         mail.getRecipients(),
                         holdFor.value());
+
+                    AuditTrail.entry()
+                        .username(session.getUsername().asString())
+                        .remoteIP(session.getRemoteAddress().getAddress().getHostAddress())
```
Comment (chibenwa) [inline server/protocols/protocols-smtp/src/main/java/org/apache/james/smtpserver/SendMailHandler.java:98]: sessionId?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/LocalDelivery.java (hunk 3 above)
Comment (quantranhong1999) [reply server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/LocalDelivery.java:76]: > Let's add a field: mailboxSessionId

Can we really get sessionId upon LocalDelivery?

File: server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remote/delivery/DeliveryRunnable.java [hunk 10]
```diff
@@ -128,6 +133,15 @@ private Mono<Void> processMail(MailQueue.MailQueueItem queueItem) {
                     // DO NOT CHANGE THIS to catch Error!
                     // For example, if there were an OutOfMemory condition caused because
                     // something else in the server was abusing memory, we would not want to start purging the retrying spool!
+                    AuditTrail.entry()
+                        .protocol("mailetcontainer")
+                        .action("RemoteDelivery")
+                        .parameters(ImmutableMap.of("mailId", mail.getName(),
+                            "mimeMessageId", mail.getMessage().getMessageID(),
+                            "sender", mail.getMaybeSender().asString(),
+                            "recipients", StringUtils.join(mail.getRecipients())))
```
Comment (chibenwa) [inline server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/remote/delivery/DeliveryRunnable.java]:
```suggestion
                        .parameters(() -> ImmutableMap.of("mailId", mail.getName(),
                            "mimeMessageId", mail.getMessage().getMessageID(),
                            "sender", mail.getMaybeSender().asString(),
                            "recipients", StringUtils.join(mail.getRecipients())))
```

Otherwize you would always join strings?

**Review by Arsnael — COMMENTED** [review]
Read it nothing to add

File: server/container/util/src/main/java/org/apache/james/util/AuditTrail.java [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+        Optional<String> userAgent = Optional.empty();
+        Optional<String> protocol = Optional.empty();
+        Optional<String> action = Optional.empty();
+        Map<String, String> parameters = ImmutableMap.of();
+
+        private Entry() {
+
+        }
+
+        public Entry username(String username) {
```
Comment (vttranlina) [inline server/container/util/src/main/java/org/apache/james/util/AuditTrail.java:51]: one more method will be less boilerplate?
`public Entry username(Username username)` 
`public Entry remoteIP(InetAddress remoteIP)`
...

Comment (quantranhong1999) [reply server/container/util/src/main/java/org/apache/james/util/AuditTrail.java:51]: > public Entry username(Username username)

I am not sure if I should add `james-core` module into `james-server-util` module. But I will try.

**Comment by Arsnael** [comment]
https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1738/5/

Related or not? I find that weird :o

**Comment by quantranhong1999** [comment]
> Related or not?

Likely related.

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
Green build.

Can you squash your fixups @quantranhong1999 ?

**Comment by quantranhong1999** [comment]
> Can you squash your fixups @quantranhong1999 ?

I am fixing some remaining review comments. Please wait ^^.

File: server/container/util/src/main/java/org/apache/james/util/AuditTrail.java (hunk 11 above)
Comment (quantranhong1999) [reply server/container/util/src/main/java/org/apache/james/util/AuditTrail.java]: > I am not sure if I should add james-core module into james-server-util module.

Circuit dependency. Sorry on this...

Comment (vttranlina) [reply server/container/util/src/main/java/org/apache/james/util/AuditTrail.java]: yes, ignored it

**Review by vttranlina — COMMENTED** [review]
What is the benefit of "Lazy evaluation" in this case?
```
.protocol(() -> "IMAP")
.action(() -> "AUTH")
```
It looks quite verbose

**Comment by quantranhong1999** [comment]
> It looks quite verbose

Likely could just be eager there.

**Review by vttranlina — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Comment by Arsnael** [comment]
The second time in a row I see this error in the builds here: https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1738/10/testReport/org.apache.james.jmap.memory.upload/InMemoryUploadServiceTest/givenQuotaExceededThenUploadShouldRepairInconsistentCurrentUsage/

I don't believe audit trail is the reason but can you double check please? But maybe more like a flaky test to me

**Comment by quantranhong1999** [comment]
> But maybe more like a flaky test to me

Yes, it is. I would prefer to investigate that flaky test later.

**Comment by Arsnael** [comment]
Same again... can you try adding https://github.com/apache/james-project/pull/1748 to see? Getting desperate here :)

**Pull request merged**